### PR TITLE
Vcert jks support

### PR DESCRIPTION
**Pull request by marcos-albornoz**
This PR is for the code to provide the support to format the output of the certificate, private key and the chain of trust to JKS similar to the behavior when the format to export is PKCS12.
In accordance of the Java Key Store specification, it's required a password to protect the Java Key Store, also it's required a password to protect the entry that will contain the certificate, private key and the chain of trust, and an alias for that entry.

So given the previous explanation, following the rules which are implemented in the current PR:

- User requests a **Java Key Store** by specifying _"jks"_ after the _-format_ switch

- User must use the _-file_ switch to specify the name of the keystore file when they specify _-format jks_.

- User specifies the password for the **store file and the key entry**( conformed by the private key, the certificate and the chain certificates) using the _-jks-password_(optionally) switch and the **key password**(_-key-password_ switch or pass phrase from prompt) respectively. If the _-jks-password_ is not provided, then the value of _key password_ is used for that.

- User always will be requested for passwords of at least 6 characters so the _-no-prompt_ switch only can be used if the _-key-password_ switch is specified.

- User must use the _-jks-alias_ switch to provide the alias for the key entry.

In summary, the rules existing for PKCS12 format applies for JKS format, with the difference that the passwords and the alias are mandatory.

Following a valid enroll declaration: 
`vcert enroll -u https://supertreat.venqa.venafi.com/ -t jhKi1xFxM5UmClnXsvaAoA== -cn amoo69.test.venafi.example.com -z devops -format jks -file test.jks -key-password abcdef -jks-alias alias1 -jks-password 123456`

### CODE REVIEW

File: cmd/vcert/result_writer.go [hunk 1]
```diff
@@ -195,6 +276,11 @@ func (r *Result) Flush() error {
 			if err != nil {
 				return fmt.Errorf("failed to encode pkcs12: %s", err)
 			}
+		} else if r.Config.Format == "jks" {
```
Comment (angelmoo) [inline cmd/vcert/result_writer.go:279]: I think as this value is being used in more than 2 places, we should convert it on a constant variable.

**Review by tr1ck3r — COMMENTED** [review]
Great work @marcos-albornoz.  I've committed some updates to error messages and help text to your branch as we discussed.  The password handling isn't quite the behavior I was expecting which is:

1. When `--format jks`, a key password is required and is specified by either `--key-password` or the password prompt when `--key-password` is omitted.
2. The key password value is always used for the **PrivateKeyEntry** password.
3. If `--jks-password` is specified, then its value is used for the **Store** password.
4. If `--jks-password` is not specified, then the key password value is used for the **Store** password.
5. The passwords must be at least 6 characters long (third-party requirement).

File: cmd/vcert/result_writer.go (hunk 1 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go]: done!

File: cmd/vcert/validators.go [hunk 2]
```diff
@@ -176,6 +170,58 @@ func validatePKCS12Flags(commandName string) error {
...
+		}
+		if (flags.csrOption == "" || flags.csrOption == "local") && flags.noPickup {
+			return fmt.Errorf(`JKS format is not allowed for the enroll or renew actions when -csr is "local" and -no-pickup is specified`)
+		}
+
+		if flags.jksPassword != "" {
+			if len(flags.jksPassword) < JKSMinPasswordLen {
+				return fmt.Errorf("Password for JKS format must be at least %d characters (see --jks-password)", JKSMinPasswordLen)
+			}
+		} else if flags.keyPassword != "" {
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply cmd/vcert/validators.go]: done!

File: cmd/vcert/result_writer.go [hunk 3]
```diff
@@ -117,6 +122,82 @@ func (o *Output) AsPKCS12(c *Config) ([]byte, error) {
...
+
+	//setting as jksPassword the value in --jks-password (if it was provided) or the value from keyPassword( --key-password or pass phrase value)
+	if c.JKSPassword != "" {
+		jksPassword = []byte(c.JKSPassword)
+	} else {
+		jksPassword = []byte(c.KeyPassword)
+	}
+
+	//adding the PK entry to the JKS
+	if err := keyStore.SetPrivateKeyEntry(c.JKSAlias, pkeIn, jksPassword); err != nil {
```
[the remark this answers is not in the document]
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go]: done!

File: cmd/vcert/result_writer.go [hunk 4]
```diff
@@ -173,24 +173,33 @@ func (o *Output) AsJKS(c *Config) ([]byte, error) {
...
 
-	var jksPassword []byte
+	var keyPass []byte
 
-	//setting as jksPassword the value in --jks-password (if it was provided) or the value from keyPassword( --key-password or pass phrase value)
+	//setting as keyPass the value from keyPassword(--key-password or pass phrase value) or the value in --jks-password
+	if c.KeyPassword != "" {
+		keyPass = []byte(c.KeyPassword)
+	} else {
+		keyPass = []byte(c.JKSPassword)
```
Comment (tr1ck3r) [inline cmd/vcert/result_writer.go:182]: We always want `keyPass = []byte(c.KeyPassword)`, never `keyPass = []byte(c.JKSPassword)` for consistency with other formats (i.e. in VCert the `--key-password` is always focused on protecting the private key).

Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:182]: mmmm… but according the chat that we had in the related email, we agreed that when it isn’t provided the private key password, that is when the --no-prompt flag is provided and the -key-password is not provided, then we can use the -jks-password value for both, the store password and also for the key entry password, and that code is for that case… or am I missing something?

File: cmd/vcert/result_writer.go [hunk 5]
```diff
@@ -117,6 +122,91 @@ func (o *Output) AsPKCS12(c *Config) ([]byte, error) {
...
+
+	//creating a PK entry
+	pkeIn := keystore.PrivateKeyEntry{
+		CreationTime: time.Now(),
+		PrivateKey:   privDER,
+		CertificateChain: []keystore.Certificate{
+			{
+				Type:    "X509",
+				Content: chainCertInBytes,
+			},
```
Comment (tr1ck3r) [inline cmd/vcert/result_writer.go:172]: I think this may be expecting one `keystore.Certificate` object per certificate in the chain (EE + all the CAs) rather than just one `keystore.Certificate` with the bytes of all of the certificates in the chain for its `Content`.

File: cmd/vcert/result_writer.go (hunk 4 above)
Comment (tr1ck3r) [reply cmd/vcert/result_writer.go:182]: Yeah, I don't want the value of the `--jks-password` ever being used for the _PrivateKeyEntry_, only for the _Store_.   The `--jks-password` is the optional and when not specified the value of the `--key-password` should be used for the _Store_ which is how that is currently coded.

Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:182]: I understand the point that you mention... the private key entry should be protected only by the key password( or the pass phrase if that was provided by prompt action) and the -jks-password should be used only to provide the password for the key store, but as we commented, what about of the case when neither the -key-password and the pass phrase were not provided( --no-prompt)? The password for the key entry is mandatory, at least it is a limit given the keystore go library which we are using

Comment (tr1ck3r) [reply cmd/vcert/result_writer.go:182]: Yes, that's right.  If a user specifies `--no-prompt` without also specifying `--key-password` with `--format jks` the expected result is an error stating passwords must be at least 6 characters long (because the default key password value is the empty string).

Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:182]: got it, in that way the user is required to provide a password for the key entry... I will do the change :-)

Comment (tr1ck3r) [reply cmd/vcert/result_writer.go:182]: Thank you @marcos-albornoz 😃

File: cmd/vcert/result_writer.go (hunk 5 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:172]: you're correct, I've update that code to add each certificate individually instead of the chain of their bytes

File: cmd/vcert/result_writer.go (hunk 4 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go]: done!

File: cmd/vcert/result_writer.go (hunk 5 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go]: done!

**Review by tr1ck3r — APPROVED** [review]
Great job @marcos-albornoz 😃 Code looks good to me now and the verifications I've done with a binary built from your branch have all been successful.

**Review by angelmoo — APPROVED** [review]
Changes looks good

File: aruba/features/format/jks.feature [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+  - User requests JKS by specifying "jks" after the -format switch
+
+  - User must use the -file switch to specify the name of the keystore file when they specify -format jks
+    (i.e. neither the -cert-file, -key-file, nor -chain-file switches may appear on the command line,
+    and console output as a base64 encoded blob will not be supported).
+
+  - User always will be request for a password of at least 6 characters when he request a JKS file so it isn't possible to use the -no-prompt switch
+
+  - User specifies the password for the JKS file and the key entry, conformed by the private kye, the certificate and the chain certificates, using, preferable,
```
Comment (rvelaVenafi) [inline aruba/features/format/jks.feature:15]: Minor typo at "private kye". It must be "private key"

File: aruba/features/format/jks.feature [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    the -jks-password or alternatively the -key-password switch
+
+  - User must use the -jks-alias switch to provide the alias for the jks entry that will conformed by the private kye, the certificate and the chain certificates
```
Comment (rvelaVenafi) [inline aruba/features/format/jks.feature:18]: Minor typo at "jks entry that will conformed by"...it must be "jks entry that will be conformed by "

Comment (rvelaVenafi) [reply aruba/features/format/jks.feature:18]: Also typo o "kye" word. Must be "key"

File: aruba/features/format/jks.feature [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    When I enroll random certificate using <endpoint> with -no-prompt -no-pickup -csr service
+    And I retrieve the certificate using <endpoint> using the same Pickup ID with -timeout 180 -key-password newPassw0rd! -file all.jks -format jks -jks-alias abc
+    And "all.jks" should be JKS archive with password "newPassw0rd!"
+    Examples:
+      | endpoint  |
+      | test-mode |
+      | TPP       |
+      # | Cloud     | # -csr service is not supported by Cloud
+
+#  Scenario Outline: Pickup JKS with typing pass phrases
```
Comment (rvelaVenafi) [inline aruba/features/format/jks.feature:146]: If these code is no longer needed, why not remove it?

File: cmd/vcert/result_writer.go [hunk 9]
```diff
@@ -117,6 +122,106 @@ func (o *Output) AsPKCS12(c *Config) ([]byte, error) {
 	return bytes, nil
 }
 
+func (o *Output) AsJKS(c *Config) ([]byte, error) {
+
+	var err interface{}
```
Comment (rvelaVenafi) [inline cmd/vcert/result_writer.go:127]: you dont need to declare err here. You can do it inline as 
var, err := myCode()

File: cmd/vcert/result_writer.go [hunk 10]
```diff
@@ -117,6 +122,106 @@ func (o *Output) AsPKCS12(c *Config) ([]byte, error) {
...
+			Content: crt.Bytes,
+		})
+	}
+
+	// getting the bytes of the PK
+	p, _ = pem.Decode([]byte(o.PrivateKey))
+	if p == nil {
+		return nil, fmt.Errorf("missing private key PEM")
+	}
+	var privDER []byte
```
Comment (rvelaVenafi) [inline cmd/vcert/result_writer.go:162]: the same with privDER, declaration not needed

File: cmd/vcert/result_writer.go [hunk 11]
```diff
@@ -117,6 +122,106 @@ func (o *Output) AsPKCS12(c *Config) ([]byte, error) {
...
+	//decrypting the PK because due the restriction that always will be requested the key password
+	//to the user(--key-password or pass phrase value from prompt) for jks format then the PK always
+	//will be encrypted with the key password provided
+	privDER, err = x509.DecryptPEMBlock(p, []byte(c.KeyPassword))
+	if err != nil {
+		return nil, fmt.Errorf("private key PEM decryption error: %s", err)
+	}
+
+	//Unmarshalling the PK
+	var privKey interface{}
```
Comment (rvelaVenafi) [inline cmd/vcert/result_writer.go:173]: The same with privKey, declaration not needed

File: cmd/vcert/utils.go [hunk 12]
```diff
@@ -44,6 +44,8 @@ const (
 	/* #nosec */
 	vCertApiKey      = "VCERT_APIKEY"
 	vCertTrustBundle = "VCERT_TRUST_BUNDLE"
+
+	JKSFormat = "jks"
```
Comment (rvelaVenafi) [inline cmd/vcert/utils.go:48]: shouldn't this be a constant?

**Review by rvelaVenafi — CHANGES_REQUESTED** [review]

File: aruba/features/format/jks.feature (hunk 8 above)
Comment (marcos-albornoz) [reply aruba/features/format/jks.feature:146]: fixing

File: aruba/features/format/jks.feature (hunk 6 above)
Comment (marcos-albornoz) [reply aruba/features/format/jks.feature:15]: fixing

File: aruba/features/format/jks.feature (hunk 7 above)
Comment (marcos-albornoz) [reply aruba/features/format/jks.feature:18]: fixing

File: cmd/vcert/result_writer.go (hunk 9 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:127]: yes, that declaration is a legancy for the previous changes

File: cmd/vcert/result_writer.go (hunk 10 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:162]: the same answer as the previous one

File: cmd/vcert/result_writer.go (hunk 11 above)
Comment (marcos-albornoz) [reply cmd/vcert/result_writer.go:173]: the same answer as the previous one :-P

File: cmd/vcert/utils.go (hunk 12 above)
Comment (marcos-albornoz) [reply cmd/vcert/utils.go:48]: ahhh... but it is in a const block

Comment (rvelaVenafi) [reply cmd/vcert/utils.go:48]: My bad, it is indeed in a constant block. All good

**Review by rvelaVenafi — APPROVED** [review]
